numpy/core/tests/test_umath.py

```diff
@@ -4023,18 +4024,28 @@ def test_float(self):
         assert_raises(TypeError, np.gcd, 0.3, 0.4)
         assert_raises(TypeError, np.lcm, 0.3, 0.4)
 
-    def test_builtin_long(self):
-        # sanity check that array coercion is alright for builtin longs
-        assert_equal(np.array(2**200).item(), 2**200)
+    def test_huge_integers(self):
+        # As of now, you can still convert to an array first and then
+        # call a ufunc with huge integers:
+        assert_equal(np.array(2**200), 2**200)
+        # but promotion rules mean that we try to use a normal integer
+        # in the following case:
+        with pytest.raises(OverflowError):
+            np.equal(2**200, 2**200)
```

Comment:
Hmmmmmm, some thoughts:
* For comparisons, yes, I think we can just allow this by saying that `PyInt + PyInt -> bool` (i.e. a comparison loop) can just use `object + object -> bool`.  For float or complex it doesn't really matter anyway (well we currently allow complex `<`, but whatever).
* Beyond that, I think things may get tricky:
  - Some functions behave differently for objects or may not have an object loop at all, so I am not sure you can do this choice universally.
  - I might start wondering if we should do this more generally, which then diverges a bit into the whole problem of returning scalars vs. arrays.

Let's focus on the first part (comparisons).  I wonder if it would be very useful if we don't also do the general `int64(2) < 2**200` implementation/hack, but I suppose it may remove some pressure at least from test suits that use NumPy asserts all over.

Reply:
I think it makes sense to remove pain points in order of simplicity!  Certainly, deferring `PyInt, PyInt` to `O,O` for comparisons seems logical. Let's move out further discussion of comparisons to the main thread.